### PR TITLE
Refactor `GENConv` linear layers for lazy init and fix doc for bipartite graphs

### PR DESCRIPTION
**Pull request by lightaime**
This PR fixes the linear layers for lazy init and doc of `GENConv` for bipartite graphs.

### CODE REVIEW

File: torch_geometric/nn/conv/gen_conv.py [hunk 1]
```diff
@@ -143,15 +154,29 @@ def __init__(
...
+            self.lin_l = Linear(aggr_out_channels, out_channels, bias=bias)
+
+        if in_channels[1] != out_channels:
+            self.lin_r = Linear(in_channels[1], out_channels, bias=bias)
+
+        channels = [out_channels]
         for i in range(num_layers - 1):
             channels.append(out_channels * 2)
         channels.append(out_channels)
         self.mlp = MLP(channels, norm=norm)
```
Comment (wsad1) [inline torch_geometric/nn/conv/gen_conv.py:178]: `MLP` also has a `bias` argument, not sure if we want to use it though.

File: torch_geometric/nn/conv/gen_conv.py [hunk 2]
```diff
@@ -143,15 +154,29 @@ def __init__(
         if isinstance(in_channels, int):
             in_channels = (in_channels, in_channels)
 
-        channels = [in_channels[0]]
+        if in_channels[0] != out_channels:
+            self.lin_n = Linear(in_channels[0], out_channels, bias=bias)
```
Comment (wsad1) [inline torch_geometric/nn/conv/gen_conv.py:158]: call it `lin_src` for readability.

File: torch_geometric/nn/conv/gen_conv.py [hunk 3]
```diff
@@ -143,15 +154,29 @@ def __init__(
...
+
+        if edge_dim is not None and edge_dim != out_channels:
+            self.lin_e = Linear(edge_dim, out_channels, bias=bias)
```
Comment (wsad1) [inline torch_geometric/nn/conv/gen_conv.py:161]:
```suggestion
            self.lin_edge = Linear(edge_dim, out_channels, bias=bias)
```

File: torch_geometric/nn/conv/gen_conv.py [hunk 4]
```diff
@@ -143,15 +154,29 @@ def __init__(
...
+
+        if isinstance(self.aggr_module, MultiAggregation):
+            aggr_out_channels = self.aggr_module.get_out_channels(out_channels)
+        else:
+            aggr_out_channels = out_channels
+
+        if aggr_out_channels != out_channels:
+            self.lin_l = Linear(aggr_out_channels, out_channels, bias=bias)
```
Comment (wsad1) [inline torch_geometric/nn/conv/gen_conv.py:169]:
```suggestion
            self.lin_aggr = Linear(aggr_out_channels, out_channels, bias=bias)
```
or maybe even `lin_aggr_out` if thats not too verbose.

File: torch_geometric/nn/conv/gen_conv.py [hunk 5]
```diff
@@ -143,15 +154,29 @@ def __init__(
...
+
+        if in_channels[1] != out_channels:
+            self.lin_r = Linear(in_channels[1], out_channels, bias=bias)
```
Comment (wsad1) [inline torch_geometric/nn/conv/gen_conv.py:172]:
```suggestion
            self.lin_dst = Linear(in_channels[1], out_channels, bias=bias)
```

**Review by wsad1 — APPROVED** [review]
Thanks, looks good. left some minor comments.

File: torch_geometric/nn/conv/gen_conv.py (hunk 1 above)
Comment (lightaime) [reply torch_geometric/nn/conv/gen_conv.py:178]: Thanks! Good point. But for the linear layers, the biases are set to `False` by default while it is `True` for the MLP. We can also set them to `True` by default if that makes it cleaner.

File: torch_geometric/nn/conv/gen_conv.py (hunk 5 above)
Comment (lightaime) [reply torch_geometric/nn/conv/gen_conv.py:172]: For the name convention, I followed what is used in other Conv layers like `SAGEConv` and `GATv2Conv`. But happy to change it if that improves readability.

**Review by rusty1s — APPROVED** [review]
Thank you! This is pretty cool.

File: test/nn/conv/test_gen_conv.py [hunk 6]
```diff
@@ -88,6 +89,20 @@ def test_gen_conv(aggr):
     assert conv((x1, x2), adj.t()).tolist() == out1.tolist()
     assert conv((x1, None), adj.t()).tolist() == out2.tolist()
 
+    value = torch.randn(row.size(0), 4)
+    adj = SparseTensor(row=row, col=col, value=value, sparse_sizes=(4, 2))
+    conv = GENConv((-1, -1), 32, aggr, edge_dim=-1)
+    conv.reset_parameters()
```
[the remark this answers is not in the document]
Comment (lightaime) [reply test/nn/conv/test_gen_conv.py]: Yes, it is just for code coverage. It is not necessary. I removed it.

**Pull request merged**